code/datums/components/companion.dm

```diff
@@ -0,0 +1,165 @@
+/// Companion module
+/datum/component/companion
+	///The mob this mob is following
+	var/mob/mob_master
+	///List of actions performed upon hearing certain things from mob_master
+	var/static/list/on_hear_behaviours = list(
+		"help" = .proc/help,
+		"you will be called" = .proc/update_name,
+		"repeat after me" = .proc/repeat_speech,
+		"goodbye" = .proc/goodbye
+	)
+	var/static/commands_info
+
+/datum/component/companion/Initialize()
+	if(!ismob(parent))
+		return COMPONENT_INCOMPATIBLE
+	if(!commands_info)
+		var/info = "Available commands:<br>"
+		for(var/key in on_hear_behaviours)
+			info += "[key]<br>"
+		commands_info = info
+
+/datum/component/companion/Destroy(force, silent)
+	unassign_mob_master()
+	return ..()
+
+/datum/component/companion/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_MASTER_MOB_CHANGE, .proc/handle_master_change)
+
+/datum/component/companion/UnregisterFromParent()
+	. = ..()
+	unassign_mob_master()
+	UnregisterSignal(parent, COMSIG_MASTER_MOB_CHANGE)
+
+///The mob gives a note to the mob_master containing available commands
+/datum/component/companion/proc/help()
+	var/mob/mob_parent = parent
+	if(!mob_parent.CanReach(mob_master))
+		say("Come closer.")
+		return
+
+	var/obj/item/paper/brassnote/new_note = new
+	new_note.info = commands_info
+	if(mob_master.put_in_any_hand_if_possible(new_note))
+		RegisterSignal(new_note, COMSIG_ITEM_DROPPED, .proc/clear_note)
+		emote("passes note")
+		return
+	else
+		qdel(new_note)
+		say("Your hands are full.")
+
+///Deletes the created commands note
+/datum/component/companion/proc/clear_note(datum/source)
+	SIGNAL_HANDLER
+	emote("...")
+	qdel(source)
+
+///Handles changing the AI's mob_master
+/datum/component/companion/proc/handle_master_change(datum/source, atom/movable/new_mob_master)
+	SIGNAL_HANDLER
+	if(mob_master == new_mob_master)
+		return
+
+	if(!new_mob_master)
+		unassign_mob_master("Farewell [mob_master]...")
+		return
+
+	var/master_update_message = "I shall follow you [new_mob_master]."
+	if(mob_master)
+		unassign_mob_master()
+		master_update_message = "Farewell [mob_master]. I will now follow [new_mob_master]."
+
+	assign_mob_master(new_mob_master, master_update_message)
+
+///Handles assigning a new master mob
+/datum/component/companion/proc/assign_mob_master(atom/movable/new_mob_master, hello_message)
+	mob_master = new_mob_master
+	RegisterSignal(parent, COMSIG_MOVABLE_HEAR, .proc/handle_mob_master_speech)
```

Comment:
missing registering in case the parent deletes

Reply:
added